miriway.cpp

```diff
@@ -68,24 +71,70 @@ std::map<std::string, ShellCommands::CmdFunctor> const wm_command =
 // Build a list of shell components from "<commands>" values and launch them all after startup
 struct ShellComponents
 {
-    explicit ShellComponents(std::function<void(std::vector<std::string> const& command_line)> launch) :
+    explicit ShellComponents(std::function<pid_t(std::vector<std::string> const& command_line)> launch) :
         launch{std::move(launch)}{}
 
     void populate(std::vector<std::string> const& cmds)
     {
         std::transform(begin(cmds), end(cmds), back_inserter(commands), ExternalClientLauncher::split_command);
     }
 
-    void launch_all() const
+    void populate_auto_restart(std::vector<std::string> const& cmds)
     {
+        std::transform(begin(cmds), end(cmds), back_inserter(auto_restart_commands), ExternalClientLauncher::split_command);
+    }
+
+    static void run_process_restart_thread(
+        bool& is_running,
+        std::vector<std::vector<std::string>> auto_restart_commands,
+        std::vector<pid_t> watched_process_list,
+        std::function<pid_t(std::vector<std::string> const& command_line)> launch)
+    {
+        while (is_running)
+        {
+            for (size_t i = 0; i < watched_process_list.size(); i++)
+            {
+                auto pid = watched_process_list[i];
+                if (kill(pid, 0) != 0)
+                {
+                    // Process does not exist
+                    watched_process_list[i] = launch(auto_restart_commands[i]);
+                }
+            }
+
+            std::this_thread::sleep_for(100ms);
+        }
+    }
+
+    void run()
+    {
+        is_running = true;
         for (auto const& command : commands)
         {
             launch(command);
         }
+
+        std::vector<pid_t> watched_process_list;
+        for (auto const& command : auto_restart_commands)
+        {
+            watched_process_list.push_back(launch(command));
+        }
+
+        process_restart_thread = std::thread(
+            run_process_restart_thread, std::ref(is_running), auto_restart_commands, watched_process_list, launch);
+    }
+
+    void stop()
+    {
+        is_running = false;
+        process_restart_thread.join();
     }
 private:
-    std::function<void (std::vector<std::string> const& command_line)> const launch;
+    std::function<int (std::vector<std::string> const& command_line)> const launch;
     std::vector<std::vector<std::string>> commands;
+    std::vector<std::vector<std::string>> auto_restart_commands;
+    bool is_running = false;
```

Comment:
This is accessed on multiple threads - could be `atomic`?

Reply:
Actually, having read the code. This should be guarded by a mutex/condition_variable combination.